qiskit/transpiler/passes/optimization/collect_multiqubit_blocks.py

```diff
@@ -0,0 +1,212 @@
+                # in the gate would fit within our desired max size
+                c_tops = set()
+                for bit in cur_qubits:
+                    c_tops.add(self.find_set(bit))
+                tot_size = 0
+                for group in c_tops:
+                    tot_size += len(self.bit_groups[group])
+                if tot_size > self.max_block_size:
+                    makes_too_big = True
+
+            if not can_process:
+                # resolve the case where we cannot process this node
+                for bit in cur_qubits:
+                    # create a gate out of me
+                    bit = self.find_set(bit)
+                    if len(self.gate_groups[bit]) == 0:
+                        continue
+                    block_list.append(self.gate_groups[bit][:])
+                    cur_set = set(self.bit_groups[bit])
+                    for v in cur_set:
+                        # reset this bit
+                        self.parent[v] = v
+                        self.bit_groups[v] = [v]
+                        self.gate_groups[v] = []
+
+            if makes_too_big:
+                # adding in all of the new qubits would make the group too big
+                # we must block off sub portions of the groups until the new
+                # group would no longer be too big
+                savings = {}
+                groups_seen = set()
+                tot_size = 0
+                for bit in cur_qubits:
+                    top = self.find_set(bit)
+                    if top in groups_seen:
+                        savings[top] = savings[top] - 1
+                    else:
+                        groups_seen.add(top)
+                        savings[top] = len(self.bit_groups[top]) - 1
+                        tot_size += len(self.bit_groups[top])
+                slist = []
+                for item in groups_seen:
+                    slist.append((savings[item], item))
+                slist.sort(reverse=True)
+                savings_need = tot_size - self.max_block_size
+                for item in slist:
+                    # remove groups until the size created would be acceptable
+                    # start with blocking out the group that would decrease
+                    # the new size the most. This heuristic for which blocks we
+                    # create does not necessarily give the optimal blocking. Other
+                    # heuristics may be worth considering
+                    if savings_need > 0:
+                        savings_need = savings_need - item[0]
+                        if len(self.gate_groups[item[1]]) >= 1:
+                            block_list.append(self.gate_groups[item[1]][:])
+                        cur_set = set(self.bit_groups[item[1]])
+                        for v in cur_set:
+                            self.parent[v] = v
+                            self.bit_groups[v] = [v]
+                            self.gate_groups[v] = []
```

Comment:
The algorithm is sub-optimal in this case. The current algorithm just uses a heuristic to top sort the gates on number of qubits they act on. To make the algorithm work here, you would have to use dynamic keys in the topological sort (e.g. keys that depend on the current blocks created). This can be done without too much struggle but takes a lot more time to run than the current pass.